// @nrwl/angular/src/generators/library/library.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.libraryGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const ngcli_adapter_1 = require("@nrwl/devkit/ngcli-adapter");
const jest_1 = require("@nrwl/jest");
const linter_1 = require("@nrwl/linter");
const init_1 = require("../../generators/init/init");
const add_linting_1 = require("../add-linting/add-linting");
const karma_project_1 = require("../karma-project/karma-project");
const add_module_1 = require("./lib/add-module");
const normalize_options_1 = require("./lib/normalize-options");
const update_lib_package_npm_scope_1 = require("./lib/update-lib-package-npm-scope");
const update_project_1 = require("./lib/update-project");
const update_tsconfig_1 = require("./lib/update-tsconfig");
const enable_strict_type_checking_1 = require("./lib/enable-strict-type-checking");
const workspace_1 = require("@nrwl/workspace");
function libraryGenerator(host, schema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        // Do some validation checks
        const options = normalize_options_1.normalizeOptions(host, schema);
        if (!options.routing && options.lazy) {
            throw new Error(`To use --lazy option, --routing must also be set.`);
        }
        if (options.enableIvy === true && !options.buildable) {
            throw new Error('enableIvy must only be used with buildable.');
        }
        if (options.publishable === true && !options.importPath) {
            throw new Error(`For publishable libs you have to provide a proper "--importPath" which needs to be a valid npm package name (e.g. my-awesome-lib or @myorg/my-lib)`);
        }
        yield init_1.default(host, Object.assign(Object.assign({}, options), { skipFormat: true }));
        const runAngularLibrarySchematic = ngcli_adapter_1.wrapAngularDevkitSchematic('@schematics/angular', 'library');
        yield runAngularLibrarySchematic(host, {
            name: options.name,
            prefix: options.prefix,
            entryFile: 'index',
            skipPackageJson: !(options.publishable || options.buildable),
            skipTsConfig: true,
        });
        devkit_1.moveFilesToNewDirectory(host, options.name, options.projectRoot);
        yield update_project_1.updateProject(host, options);
        update_tsconfig_1.updateTsConfig(host, options);
        yield addUnitTestRunner(host, options);
        updateNpmScopeIfBuildableOrPublishable(host, options);
        add_module_1.addModule(host, options);
        setStrictMode(host, options);
        yield addLinting(host, options);
        if (options.standaloneConfig ||
            devkit_1.getWorkspaceLayout(host).standaloneAsDefault) {
            yield workspace_1.convertToNxProjectGenerator(host, {
                project: options.name,
                all: false,
            });
        }
        if (!options.skipFormat) {
            yield devkit_1.formatFiles(host);
        }
        return () => {
            devkit_1.installPackagesTask(host);
        };
    });
}
exports.libraryGenerator = libraryGenerator;
function addUnitTestRunner(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (options.unitTestRunner === 'jest') {
            yield jest_1.jestProjectGenerator(host, {
                project: options.name,
                setupFile: 'angular',
                supportTsx: false,
                skipSerializers: false,
            });
        }
        else if (options.unitTestRunner === 'karma') {
            yield karma_project_1.default(host, {
                project: options.name,
            });
        }
    });
}
function updateNpmScopeIfBuildableOrPublishable(host, options) {
    if (options.buildable || options.publishable) {
        update_lib_package_npm_scope_1.updateLibPackageNpmScope(host, options);
    }
}
function setStrictMode(host, options) {
    if (options.strict) {
        enable_strict_type_checking_1.enableStrictTypeChecking(host, options);
    }
    else {
        enable_strict_type_checking_1.setLibraryStrictDefault(host, options.strict);
    }
}
function addLinting(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (options.linter === linter_1.Linter.None) {
            return;
        }
        yield add_linting_1.default(host, {
            projectName: options.name,
            projectRoot: options.projectRoot,
            prefix: options.prefix,
        });
    });
}
exports.default = libraryGenerator;
